<a href="https://colab.research.google.com/github/insancs/pca-python/blob/main/pca.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn import datasets
from sklearn.model_selection import train_test_split

In [ ]:
# Load iris dataset
iris = datasets.load_iris()

# Separate attribute and label
atribut = iris.data
label = iris.target
iris

{'DESCR': '.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ===========

In [ ]:
# separate training data and test data
X_train, X_test, y_train, y_test = train_test_split(atribut, label, test_size=0.2)

Kita akan menggunakan model Decision Tree dan menghitung berapa akurasinya tanpa menggunakan PCA.

In [ ]:
from sklearn import tree

# Call Decision Tree model
decision_tree = tree.DecisionTreeClassifier()
first_model = decision_tree.fit(X_train, y_train)
# Print accuracy
first_model.score(X_test, y_test)

0.9

Akurasi tanpa PCA adalah 0.9 atau 90%.
Kemudian kita akan menggunakan PCA dan menghitung variance dari setiap atribut.

In [ ]:
from sklearn.decomposition import PCA

# create a PCA object with 4 principal components
pca = PCA(n_components=4)

# PCA implementation on dataset
pca_attributes = pca.fit_transform(X_train)

# see the variance of each attribute
pca.explained_variance_ratio_

array([0.931342  , 0.04688987, 0.01605043, 0.0057177 ])

Hasilnya adalah 1 atribut memiliki variance sebesar 0.931, yang berarti atribut tersebut menyimpan informasi yang tinggi dan jauh lebih signifikan dari atribut lain.

Melihat dari variance sebelumnya kita bisa mengambil 2 principal component terbaik karena total variance nya adalah 0.977 yang sudah cukup tinggi.

In [ ]:
# PCA with 2 principal components
pca2 = PCA(n_components = 2)
X_train_pca = pca2.fit_transform(X_train)
X_test_pca = pca2.fit_transform(X_test)

In [ ]:
# Testing accuracy
model2 = decision_tree.fit(X_train_pca, y_train)
model2.score(X_test_pca, y_test)

0.8

Dari percobaan di atas bisa kita lihat bahwa dengan hanya 2 principal component atau 2 atribut saja model masih memiliki akurasi yang cukup tinggi, yaitu 80%. Dengan principal component kamu bisa mengurangi atribut yang kurang signifikan dalam prediksi dan mempercepat waktu pelatihan sebuah model machine learning.